### PR TITLE
Compute lifetimes in scope at diagnostic time

### PR DESCRIPTION
**Pull request by cjgillot**
The set of available lifetimes is currently computed during lifetime resolution on HIR.  It is only used for one diagnostic.

In this PR, HIR lifetime resolution just reports whether elided lifetimes are well-defined at the place of use.  The diagnostic code is responsible for building a list of lifetime names if elision is not allowed.

This will allow to remove lifetime resolution on HIR eventually.

### CODE REVIEW

**Comment by cjgillot** [comment]
I wonder if we shouldn't just remove this query and always suggest named lifetimes.

**Comment by cjgillot** [comment]
I added 2 commits to remove the `lifetime_scope_map`.  Having a dedicated query just for this diagnostic is IMO not worth the complexity.  This is replaced by a specific suggestion for statics and consts to add `'static` lifetime where appropriate.

**Comment by compiler-errors** [comment]
This looks good.

@bors r+

File: src/test/ui/suggestions/missing-lifetime-specifier.stderr [hunk 1]
```diff
@@ -171,8 +171,8 @@ LL | pub union Qux<'t, 'k, I> {
    |           ^^^ --  --
 help: add missing lifetime argument
    |
-LL |     static e: RefCell<HashMap<i32, Vec<Vec<Qux<'static, '_, i32>>>>> = RefCell::new(HashMap::new());
-   |                                                       ++++
+LL |     static e: RefCell<HashMap<i32, Vec<Vec<Qux<'static, 'static, i32>>>>> = RefCell::new(HashMap::new());
+   |                                                       +++++++++
```
Comment (klensy) [inline src/test/ui/suggestions/missing-lifetime-specifier.stderr:175]: Spans looks slightly bigger than needed, but that was before that PR.

Comment (compiler-errors) [reply src/test/ui/suggestions/missing-lifetime-specifier.stderr:175]: What do you mean spans look larger? This is a suggestion, the fixup is to add `, 'static`.

Comment (klensy) [reply src/test/ui/suggestions/missing-lifetime-specifier.stderr:175]: Ah, that is span to added suggestion, not old span, you right.

**Pull request merged**